File: 05-binary_search/00875-Koko_Eating_Bananas/00875-koko_eating_bananas.java
// 875: Koko Eating Bananas
// https://leetcode.com/problems/koko-eating-bananas/

import java.util.Arrays;

class Solution {
    // SOLUTION
    public int minEatingSpeed(int[] piles, int h) {
        int l = 1, r = Arrays.stream(piles).max().getAsInt();
        
        while (l < r) {
            int total = 0;
            int m = (l + r) / 2;

            for (int p : piles)
                total += (p + m - 1) / m;
            if (total > h)
                l = m + 1;
            else
                r = m;
        }

        return l;
    }


    public static void main(String[] args) {
        Solution o = new Solution();
        
        // INPUT
        int[] piles = {3,6,7,11};
        int h = 8;

        // OUTPUT
        var result = o.minEatingSpeed(piles, h);
        System.out.println(result);
    }
}

File: 05-binary_search/00875-Koko_Eating_Bananas/00875-koko_eating_bananas.py
# 875: Koko Eating Bananas
# https://leetcode.com/problems/koko-eating-bananas/


class Solution:
    # SOLUTION
    def minEatingSpeed(self, piles: list[int], h: int) -> int:
        l, r = 1, max(piles)
        
        while (l < r):
            total = 0
            m = (l + r) // 2

            for p in piles:
                total += (p + m - 1) // m
            if (total > h):
                l = m + 1
            else:
                r = m

        return l



if __name__ == "__main__":
    o = Solution()

    # INPUT
    piles: list[int] = [3,6,7,11]
    h: int = 8

    # OUTPUT
    result = o.minEatingSpeed(piles, h)
    print(result)